traits/observers/_set_item_observer.py

```diff
@@ -0,0 +1,203 @@
+            raise ValueError(
+                "Expected a TraitSet to be observed, "
+                "got {!r} (type: {!r})".format(object, type(object))
+            )
+
+        yield object
+
+    def iter_objects(self, object):
+        """ Yield the content of the set if the given object is an observable
+        set. Otherwise, raise an error, unless the observer is optional.
+
+        The content of the set will be passed onto the children observer(s)
+        following this one in an ObserverGraph.
+
+        Parameters
+        ----------
+        object: object
+            Object provided by another observers or by the user.
+
+        Yields
+        ------
+        value : object
+
+        Raises
+        ------
+        ValueError
+            If the given object is not an observable set and optional is false.
+        """
+        if not isinstance(object, TraitSet):
+            if self.optional:
+                return
+            raise ValueError(
+                "Expected a TraitSet to be observed, "
+                "got {!r} (type: {!r})".format(object, type(object))
+            )
+
+        yield from object
+
+    def get_notifier(self, handler, target, dispatcher):
+        """ Return a notifier for calling the user handler with the change
+        event.
+
+        Returns
+        -------
+        notifier : TraitEventNotifier
+        """
+        return TraitEventNotifier(
+            handler=handler,
+            target=target,
+            dispatcher=dispatcher,
+            event_factory=set_event_factory,
+            prevent_event=lambda event: False,
+        )
+
+    def get_maintainer(self, graph, handler, target, dispatcher):
+        """ Return a notifier for maintaining downstream observers when
+        a set is mutated.
+
+        Parameters
+        ----------
+        graph : ObserverGraph
+            Description for the *downstream* observers, i.e. excluding self.
+        handler : callable
+            User handler.
+        target : object
+            Object seen by the user as the owner of the observer.
+        dispatcher : callable
+            Callable for dispatching the handler.
+
+        Returns
+        -------
+        notifier : ObserverChangeNotifier
+        """
+        return ObserverChangeNotifier(
+            observer_handler=_observer_change_handler,
+            event_factory=set_event_factory,
+            prevent_event=lambda event: False,
+            graph=graph,
+            handler=handler,
+            target=target,
+            dispatcher=dispatcher,
+        )
+
+    def iter_extra_graphs(self, graph):
+        """ Yield new ObserverGraph to be contributed by this observer.
+
+        Parameters
+        ----------
+        graph : ObserverGraph
+            The graph this observer is part of.
+
+        Yields
+        ------
+        ObserverGraph
+        """
+        # Unlike CTrait, no need to handle trait_added
+        yield from ()
```

Comment:
That's a neat trick for an empty generator. I've always used horrible things like `if False: yield None`, but this is better. (It's always felt like a violation of EIBTI that a Python generator function is _implicitly_ a generator function because it uses `yield` somewhere, rather than because it's declared to be a generator function somehow.)